File: leetcode/CPP/1D-Dymanic-Programming/0005-Longest-Palindromic-Substring.cpp
/*
    0005. Longest Panlindromic Substring 

    Given a string s, return longest palindromic sbstirng in s.

    Example 1:

    Input: s = "babad"
    Output: "bab"
    Explanation: "aba" is also a valid answer.
*/

#include <iostream>
#include <string>
#include <vector>

class DynamicProgramming1D {
    public:
        // Expand Around the Center
        std::string longestPalindromeCenter(std::string s) {
            if (s.empty()) return "";
            int start = 0, end = 0;

            auto expandAroundCenter = [&](int left, int right) {
                // Expand outwards as long as the characters at the left and right indices are the same
                while (left >= 0 && right < s.size() && s[left] == s[right]) {
                    left--;         // Move left pointer to the left
                    right++;        // Move the right pointer to the right 
                }
                return right - left - 1;  // Length of the palindrome 
            };

            for (int i = 0; i < s.size(); i++) {
                // Check for odd-length palindromes with center at i
                int len1 = expandAroundCenter(i, i);       
                // Check for even-length palindromes with centers at i and i + 1
                int len2 = expandAroundCenter(i, i + 1);   
                int len = std::max(len1, len2);
                // If the found palindrome is longer than the previously found ones, update start and end 
                if (len > end - start) {
                    // i center of the current palindrome
                    // (len - 1) / 2: Calculates the distance from the center i to the start of the palindrome
                    start = i - (len - 1) /2;
                    // i center of palindrome 
                    // len / 2: Calculates the distance from the center i to the end of the palindrome.
                    end = i + len / 2;
                }
            }
            return s.substr(start, end - start + 1);
        }

        // Dynamic Programming 
        std::string longestPalindromeDP(std::string s) {
            if (s.length() <= 1) return s;

            int maxLength = 1;  
            int startIndex = 0;
            int endIndex = 0;
            std::vector<std::vector<bool>> isPalindrome(s.length(), std::vector<bool>(s.length(), false));

            for (int i = 0; i < s.length(); i++) {
                isPalindrome[i][i] = true;      // Single character palindromes 

                // Check for palindromic substrings ending at index i
                for(int j = 0; j <i; ++j) {
                    // s[j] == s[i] means the characters at index j and i are the same 
                    // (i - j <= 2) means the substring is either two characters or three characters with 
                    // middle one already a palindrome
                    // isPalindrome[j + 1][i - 1] means the substring inside the current bounds is already 
                    // a palindrome  
                    if (s[j] == s[i] && (i - j <= 2 || isPalindrome[j + 1][i - 1])) {
                        isPalindrome[j][i] = true;
                        if (i - j + 1 > maxLength) {
                            maxLength = i - j + 1;
                            startIndex = j; 
                            endIndex = i;
                        }
                    }
                }
            }
            return s.substr(startIndex, endIndex - startIndex + 1);
        }
};


int main () {
    std::string s = "babad";
    DynamicProgramming1D dynamic;
    
    std::string result = dynamic.longestPalindromeDP(s);

    std::cout << "Longest Palindromic Substring : " << result << std::endl;

    return 0;
}

File: leetcode/OCaml/1D-Dynamic-Programming-OCaml/0005-Longest-Palindromic-Substring.ml
(* 
  0005. Longest Palindromic Substring (Medium)
  
  Given a string s, return the longest palindromic substring in s.

  Example 1:

  Input: s = "babad"
  Output: "bab"
  Explanation: "aba" is also a valid answer.
*)

let longest_palindrome_center s =
  let len = String.length s in 
  if len = 0 then "" 
  else 
    let expand_around_center left right = 
      let rec expand left right = 
        if left >= 0 && right < len && s.[left] = s.[right] then 
          expand (left - 1) (right + 1)
        else
          right - left - 1
        in
        expand left right 
      in
      
      let rec find_longest start end_ i = 
        if i >= len then String.sub s start (end_ - start + 1)
        else
          let len1 = expand_around_center i i in 
          let len2 = expand_around_center i (i + 1) in 
          let len = max len1 len2 in 
          if len > end_ - start then 
            let new_start = i - (len - 1) / 2 in 
            let new_end = i + len / 2 in 
            find_longest new_start new_end (i + 1)
          else    
            find_longest start end_ (i + 1)
          in
          find_longest 0 0 0 


let longest_palindrome_dp s  =
  let len = String.length s in 
  if len <= 1 then s 
  else 
    let is_palindrome = Array.make_matrix len len false  in
    let max_length = ref 1 in 
    let start_index = ref 0 in
    let end_index = ref 0 in 
    
    (* All substrings of length 1 are palindromes *)
    for i = 0 to len - 1 do
      is_palindrome.(i).(i) <- true 
    done;

    (* check for substring of length greater than 1 *)
    for i = 1 to len -  1 do
      for j = 0 to i - 1 do
        if s.[j] = s.[i] && (i - j <= 2 || is_palindrome.(j + 1).(i - 1)) then 
          begin
            is_palindrome.(j).(i) <- true;
            if i - j + 1 > !max_length then 
            begin
              max_length := i - j + 1;
              start_index := j;
              end_index := i 
            end 
        end
      done
    done;
    String.sub s !start_index (!end_index - !start_index + 1)

let () = 
  let test s expected = 
    let result = longest_palindrome_center s in 
    Printf.printf "longest_palindrome_center \"%s\" = \"%s\" (expected \"%s\")\n" s result expected 
  in
    test "babad" "aba";
    test "cbbd" "bb"

let () = 
  let test s expected = 
    let result = longest_palindrome_dp s in 
    Printf.printf "longest_palindrome_dp \"%s\" = \"%s\" (expected \"%s\")\n" s result expected
  in
  test "babad" "aba";
  test "cbbd" "bb"